### ISSUE TITLE
Allow custom annotator classes to append information into the token model

### ISSUE THREAD

**Issue by Liebeck** [explosion/spaCy#1085]
**Background**: 
I finally found some time to port my German lemmatizer to Python. We discussed this previously in #390 , #846, and #1024. The port is now available [here](https://github.com/Liebeck/IWNLP-py). Now, I want to integrate it in spaCy. [Customizing the pipeline](https://spacy.io/docs/usage/customizing-pipeline) is a good start but *arbitrary_fixup_rules* is stateless and writes to existing attributes of the token class.

**Feature request**
I'm currently trying to integrate multiple custom stateful annotators into the pipeline. I'm struggling with how to write their results back into the doc and token objects.

Based on examples of stateful annotators (https://spacy.io/docs/usage/deep-learning) and another I example I can't find at the moment, I managed to hook me into the pipeline and get my lemmatization code working:

```python
from iwnlp_lemmatizer_spacy import IWNLPLemmatizerSpacy

nlp = spacy.load('de')
nlp.pipeline = [nlp.tagger, nlp.parser, nlp.matcher, nlp.entity,
			       IWNLPLemmatizerSpacy.load(lemmatizer_path='IWNLP.Lemmatizer_20170501.json')]
```

```python
from iwnlp.iwnlp_wrapper import IWNLPWrapper

class IWNLPLemmatizerSpacy(object):
    @classmethod
    def load(cls, lemmatizer_path):
        lemmatizer = IWNLPWrapper(lemmatizer_path=lemmatizer_path)
        return cls(lemmatizer)

    def __init__(self, lemmatizer):
        self._lemmatizer = lemmatizer

    def __call__(self, doc):
        for token in doc:
            lemmas = self._lemmatizer.lemmatize(token.text, pos_universal_google=token.pos_)
            # token.iwnlp_lemma = lemmas
```

My intuitive assumption was that I can simply add the values of my custom annotators directly into the token object. Unfortunately, this doesn't work and I understand this due to https://github.com/explosion/spaCy/blob/df2ac8b843da511a1f83fde4b4c64e6fe46bfdfc/spacy/tokens/token.pyx 

I'd like to create multiple custom annotators that add new values (such as multiple sentiment values, each determined by a different method) on a token level. As far as I can see, this is currently not possible. For this example I can probably set the lemma attribute but I have a couple of other annotations in mind where no corresponding attribute exists at the moment. My workaround will probably be that I don't use custom annotators directly in the pipeline and "roll my own pipeline" after spaCy is done with its pipeline. I'm grateful for any other suggestions:

This sums up to: **Please add a way to append information directly into the token object**

## Your Environment
* **spaCy version:** 1.8.2
* **Platform:** Linux-4.2.0-27-generic-x86_64-with-debian-jessie-sid
* **Python version:** 3.5.2
* **Installed models:** de

**Comment by honnibal** [comment]
Thanks for your work on this!

Unfortunately adding to the **token** specifically is very difficult: the Python `Token` object doesn't own any data --- it's basically a wrapper around a `(doc, i)` tuple, with everything exposed through properties. This lets us have a single source of truth --- the `Doc` object.

You can set data on the `Doc` object with the `doc.user_data` dictionary. However, the usage of this doesn't feel very satisfying. I very much feel that how the lines of your code look matters --- so I wouldn't find that writing doc.user_data['sentiment_polarity']` all the time to be acceptable.

So, I think there's an open design question here. The solution-space seems to me to be:

1. Go fully dynamic in the `Token` and/or `Doc` object. This would mean implementing a `__getattr__` hook, so that the user can make custom `token.sentiment_polarity` etc attributes. I've always seen `__getattr__` as a method of last resort, so I've been reluctant to do it this way.

2. Provide a way to yield user subclasses of `Doc` or `Token`. The `Doc` object could have a function `token_factory` that it calls into, instead of instantiating `Token` directly.

3. Users could wrap instead of subclass. In spaCy 2 we're changing the pipeline logic so that pipeline components return objects, which they're allowed to destroy/modify in place. But the component could also return a wrapper, instead of the object itself. This would allow you to create an extension that proxies to the native `Doc`, but intercepts particular calls.

4. Add a "user space" area that allows users to do arbitrary dynamic stuff. 

Solution 1 requires the least change to the code, but I think in the long-term it's quite damaging. If we open `__getattr__`, then every time we add a property to `Token`, we're making a potentially breaking change. We're letting all users write into the same flat namespace the library is using.

Solution 2 would work, but extensions will be default incompatible with each other, because you can only have one token factory. There will be no neat way to reconcile them --- to have two separate packages that implement their own components, you'd have to have a shared dependency. People won't do this -- we'll just have incompatible components.

Solution 3 suffers the same incompatibility problems, and wrappers are always brittle. Wrapping the `Doc` to yield wrapped `Token` instances will also be really ugly. This is the least appealing option, except that it works entirely in user-space --- the library doesn't need to know anything about it.

Solution 4 gives up the "entirely native" idea. User-space extensions will look distinct from the native functionality. I suggest:

```
doc._.my_method()
token._.polarity
```

The `Doc._` attribute would be a Python object that users can write to. You can use arbitrary Python dynamic tricks to give yourself whatever syntactic sugars you need from this. I think `Token._` should be a property that delegates to the `Doc._` object. This allows us to avoid having the `Token` own any data. `Span._` would behave similarly.

The workflow would be:

* Add component to pipeline
* Component sets an object to `Doc._`
* User writes `doc[i]._`. This generates a call to `doc._.get_token(i)`, which can return any object you like. 

What do you think?

**Comment by Liebeck** [comment]
That's a tough design decision. I needed to take some time to think about it. Maybe @ines also has some thoughts on this?

From my point of view, it depends on how you want to integrate custom annotators into your pipeline. If you take a look at ```IWNLPLemmatizerSpacy``` how do you want your end users to integrate it into their pipeline? Should everyone write such classes themselves or should there be a way that a library can instantly be hooked up, as in ```nlp.pipeline=[..., IWNLPLemmatizerSpacy.load(lemmatizer_path='IWNLP.Lemmatizer_20170501.json')]```  with the wrapper class being in imported library?

That would be a nice way to just include a 3rd party library into the pipeline. However, such annotators somehow need to save their results. How about solution 2, using a subclass of token, and additionally specifing to which property each annotator can write to?

For instance ```nlp.custom_token_class=MyCustomTokenClass``` and ```nlp.pipeline=[..., IWNLPLemmatizerSpacy.load(lemmatizer_path='IWNLP.Lemmatizer_20170501.json', output_property='iwnlp_lemma')]```. ```MyCustomTokenClass``` would inherit from Token and contain all the properties that a user would need for his NLP pipeline. This way I could add multiple properties, each annotator would know where to write its results and multiple annotators would not clash because the end user can specify where the results are written to. Is this something that would work? ;)

**Comment by ines** [comment]
Sorry for the late reply – I was holding back on this because we were hoping to already have v2.0.0 alpha ready by now. This would have made some explanations a little easier. (But unfortunately, we got stuck on some serializer issues, so it's taking a bit longer.)

In general, I do like the `_` solution for user hooks, as it keeps a clearer separation and makes it easier to ensure backwards compatibility. For example, if you've implemented your own `.coref` property and spaCy claims it one day, it'll break  your code. Similarly, as we have more and more production users with sizable code bases, this solution will also make it much easier to tell what's built-in and what's custom. Just by looking at the code, you'll immediately know that `doc.sentiment` is spaCy, and `doc._.sentiment_score` isn't.

### Custom pipelines in v2.0

In spaCy v2.0, you'll still be able to modify the pipeline and insert your own custom functions, for example `nlp.pipeline.insert(1, my_custom_component)`. A component function receives a `Doc` and returns a modified version of it, which will then be passed on to the next component in the pipeline.

![bildschirmfoto 2017-05-31 um 13 14 21](https://cloud.githubusercontent.com/assets/13643239/26629603/6165fe62-4603-11e7-92f0-ab8a45c26a4f.png)

**Stateful components** can be created and registered as "factories" that receive the shared `Vocab` and return a component function. This lets you do whatever you want within the factory:

```python
class CustomLemmatizer(object):
    def __init__(self, vocab):
        self.vocab = vocab
        self.lemmatizer = IWNLPWrapper(lemmatizer_path='IWNLP.Lemmatizer_20170501.json')

    def get_token_lemma(self, token):
        return self.lemmatizer.lemmatize(token.text, pos_universal_google=token.pos_)

    def __call__(self, doc):
        # assign lemmas to user_data
        doc.user_data[self.output_property] = [self.get_token_lemma(token) for token in doc]
        return doc

# register factory with spaCy
spacy.set_factory('custom_lemmatizer', CustomLemmatizer)

# create new Language object and pipeline with defaults + custom factory
nlp = German(pipeline=['tensorizer', 'tagger', 'parser', 'custom_lemmatizer', 'ner'])
```
On initialisation of the `Language` object, spaCy will look up each component in its factories registry. Specifying the pipeline components as strings has a few advantages: It'll let you use pipeline components without having to know/remember where they are and how to implement them. It'll also let models define their pipelines as part of their meta.json, e.g. `['tagger', 'parser']`.

So if you're writing a third-party component, you could even go one step further and write a function that takes a few arguments, like the path, and returns a factory – or creates a factory and registers it with spaCy straight away.

```python
# create factory and register it with spaCy using a custom ID
register_lemmatizer('custom_lemmatizer', path='IWNLP.Lemmatizer_20170501.json', doc_output_property='iwnlp_lemmas', token_output_property='iwnlp_lemmas')

# 'custom_lemmatizer' can now be added to a pipeline
nlp = German(pipeline=['tensorizer', 'tagger', 'parser', 'custom_lemmatizer', 'ner'])
```

The object assigned to `_` would then take care of putting this all together. Going via the `Doc` and its `user_data` is necessary here, because tokens are only views of the document and don't actually own any data themselves. The API for this isn't 100% clear yet, but you should be able to just assign getters to the `Doc`'s and `Token`'s `_` object.

```python
iwnlp_lemmas = lambda doc: doc.user_data['iwnlp_lemmas'])
iwnlp_lemma = lambda token: token.doc.user_data['iwnlp_lemmas'][self.token.i])
```

**Comment by Liebeck** [comment]
Thanks! I skimmed over parts of the new documentation and I should be able to better implement IWNLP into spaCy this week and to also adapt other pipeline components.

I know that both of you are pretty busy right now, but have you been thinking about where you want to collect and share community-made pipeline components?

**Comment by ines** [comment]
Great – let us know how you get on! Some parts of the v2 docs are still under construction, so feedback is always appreciated.

 > I know that both of you are pretty busy right now, but have you been thinking about where you want to collect and share community-made pipeline components?

I've actually been thinking about this just the other day! Of course, we don't want to make too many plans before people even start developing their extensions, but here are some ideas:

* Encourage users to create their spaCy extensions as individual packages that expose the components or factories (maybe even come up with best practices and a recommended API). This will make it easy for users to manage their dependencies, the maintainers of the packages can choose how to distribute them (GitHub, PyPi), and other users will be able to contribute.

* I don't think bundling them into a `spacy-contrib` or `spacy-extensions` package like other libraries do would be a good solution for us. Extensions are very specific, and users shouldn't have to install a bunch of stuff they don't need just to use one particular component. Similarly, I don't think we should make people submit them to an "official" repository – if someone made a spaCy extension, they should be able to showcase their work on their own GitHub profiles.

* Extend the current showcase section of the docs to be more of a directory of third-party libraries and extensions. Maybe include the directory of pipeline components on pipeline docs page. Creating an `awesome-spacy` type of repository could also be pretty cool.

* Add a usage guide on "how to create a spaCy extension for your project".

**Comment by ines** [comment]
@Liebeck We've finally done another round of iteration on the pipeline components and have ended up with a solution that should give you all the tools and functionalities you need to integrate your lemmatizer and share it as a spaCy plugin. It's also much nicer than the proposal I originally outlined in this thread. (I'm actually super excited about this one and have had a lot of fun experimenting with different kinds of spaCy extensions – so many new possibilities!)

The update will be included in the next alpha release. #1408 also has a few code examples of pipeline component extensions setting custom attributes on the `Doc`, `Token` and `Span` that you should be able to adapt for your use case 😊  (I've also written more extensive docs for the website which are already included in the PR, but not live on [alpha.spacy.io](https://alpha.spacy.io) yet.)

For reference:
#1381: Proposal: A component-based processing pipeline architecture via `Doc._`, `Token._` and Span._
#1396: Improve pipeline and factory management
#1408: Custom attributes via `Doc._`, `Token._` and `Span._`

**Comment by Liebeck** [comment]
@ines Sounds great! I'm currently in the last phase of my dissertation which is why I only skimmed over #1381. I'm pretty sure I can get my two planned extensions working with the examples you provided ;)

How was the lemmatization dictionary from #1413 created? I'm curious about how good the coverage and the accuracy of the lemmatization is on German corpora, or how it compares to my [IWNLP](https://github.com/Liebeck/IWNLP-py) and the current [results](http://www.iwnlp.com/iwnlp_results.html). Are you still interested in having my Python implementation as a spaCy extension or has that become obsolete with #1413?

A couple of months ago, we discussed how you want to share spaCy extensions among users. I agree with you that it is probably best that each extension developer has his or her own repositories where the individual code is stored.

@ines @honnibal 
However, I think it is necessary that you somewhere provide a list of available extensions on the website. Otherwise it becomes difficult to find extensions.  
I imagine something like https://spacy.io/docs/usage/showcase with a list of the extensions. Each extensions should provide a 1-2 sentence description, a link to the repository, and an example usage page in your documentation.
Would that be ok for you?

**Comment by ines** [comment]
> How was the lemmatization dictionary from #1413 created?

The current lists were contributed by @Gregory-Howard in #1024 and according to the thread, they're based on these datasets: http://www.lexiconista.com/Datasets/ We haven't run any evaluation yet, since it was just implemented, but I'd definitely be curious to see how it compares.

> Are you still interested in having my Python implementation as a spaCy extension or has that become obsolete with #1413?

Absolutely – if you're still interested of course 😉  With the built-in lookup lemmatizers, we obviously want to offer a good, general-purpose solution. But I think there will always be a need for third-party extensions that offer more specific functionality and customisation.

I think `IWNLP` would actually be a perfect example of a spaCy extension. It'd allow users to load in different data in the format you support, and you could add your own `._.iwnlp_lemma` attributes to the `Token`, or maybe even some cool methods.

Usage could then look something like this:

```python
from spacy_iwnlp import IWNLP
nlp = spacy.load('de')
component = IWNLP(nlp.vocab, path='data.json', ignore_case=True)
nlp.add_pipe(component, after='tagger')

doc = nlp("Ich bin ein zerstreuter Satz.")
doc[3]._.iwnlp_lemma  # 'zerstreut'
doc[3]._.iwnlp_all_lemmas  # ['zerstreut', 'zerstreuen']
```

> However, I think it is necessary that you somewhere provide a list of available extensions on the website. Otherwise it becomes difficult to find extensions.

👍  Yes, this is pretty much exactly what we've had in mind for the "Resources" section of the new website: https://alpha.spacy.io/usage/resources

I also wrote a little spaCy plugin for semi-advanced emoji handling (which was great for debugging and can function as the first example), plus a work-in-progress guide on best-practices for developing custom extensions ([raw markup here](https://github.com/explosion/spaCy/blob/51519251c23654a9f44f2f995b473e24735a5f2c/website/usage/_processing-pipelines/_extensions.jade), will go live on the alpha site as soon as the new `spacy-nightly` version is released).

**Comment by Liebeck** [comment]
That sound's great ;)

I'm pretty swamped right now, unfortunately. I will benchmark the new German lemmatizer in January. Hopefully, I'll find some time in the next two weeks to implement two extensions :)

**Comment by ines** [comment]
No worries, take your time! The nice thing about v2.0 still being in alpha is that it hopefully gives everyone enough time to experiment with it and implement extensions they want to build.

Just published `v2.0.0a17` btw, which includes the changes to the pipeline and attribute extensions. And here's my emoji extension 😃 https://github.com/ines/spacymoji

**Comment by benhachey** [comment]
Hi @ines - Very excited about custom attributes. I'm just testing and can't get deserialisation to work, e.g.:


```python
import spacy
from spacymoji import Emoji

nlp = spacy.load('en_core_web_sm')
emoji = Emoji(nlp)
nlp.add_pipe(emoji, first=True)

doc = nlp(u"This is a test 😻 👍🏿")
assert doc._.has_emoji == True
assert doc[2:5]._.has_emoji == True
assert doc[0]._.is_emoji == False
assert doc[4]._.is_emoji == True
assert doc[5]._.emoji_desc == u'thumbs up dark skin tone'
assert len(doc._.emoji) == 2
assert doc._.emoji[1] == (u'👍🏿', 5, u'thumbs up dark skin tone')

from spacy.tokens import Doc
from spacy.vocab import Vocab
bin_doc = doc.to_bytes()
doc2 = Doc(Vocab()).from_bytes(bin_doc)
```

Raises:
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-11-0278d6942195> in <module>()
      3 
      4 bin_doc = doc.to_bytes()
----> 5 doc2 = Doc(Vocab()).from_bytes(bin_doc)

doc.pyx in spacy.tokens.doc.Doc.from_bytes()

~/repos/ArtemisML/env2/lib/python3.6/site-packages/spacy/util.py in from_bytes(bytes_data, setters, exclude)
    474 
    475 def from_bytes(bytes_data, setters, exclude):
--> 476     msg = msgpack.loads(bytes_data, encoding='utf8')
    477     for key, setter in setters.items():
    478         if key not in exclude and key in msg:

~/repos/ArtemisML/env2/lib/python3.6/site-packages/msgpack_numpy.py in unpackb(packed, object_hook, encoding, **kwargs)
    185 
    186     kwargs['object_hook'] = object_hook
--> 187     return _unpacker.unpackb(packed, encoding=encoding, **kwargs)
    188 
    189 load = unpack

msgpack/_unpacker.pyx in msgpack._unpacker.unpackb (msgpack/_unpacker.cpp:2068)()

TypeError: unhashable type: 'list'
```

Using:
```
en-core-web-sm==2.0.0a7
spacy-nightly==2.0.0a17
spacymoji==0.0.1
```

Am I doing something wrong?

**Comment by honnibal** [comment]
@benhachey This was a bug -- I forgot that msgpack doesn't distinguish between tuples and lists. When serializing, there's no problem, but when the message was read back in, the tuple gets converted to a list, which doesn't work as a dict key.

I've changed the user data serialization to write out the keys and values separately. This was we can have `use_list=False` in the keys, while returning lists in the values. It still sucks that lists get turned into tuples in the values, but there's no easy solution.

Pickling the `Doc` should work now, and there should be no restriction on the data types within `user_data` if you're using pickle.

I'm currently pickling the `doc.user_hooks`, `doc.user_span_hooks` and `doc.user_token_hooks` fields by first encoding them with `dill`, which allows you to use lambda functions and other flexible stuff. This means these should deserialize quite reliably. The downside is that the data will be duplicated if you serialize a list of `doc` objects. This isn't great, but it's not obvious what to do about it. I suppose people can clear their hooks before pickling, and restore them afterwards. This seems better than the pickle failing.

**Comment by benhachey** [comment]
Cool, thx. Currently testing v2 on AWS Lambda. Looks good so far!

**Comment by smartinsightsfromdata** [comment]
@Liebeck Any news on integrating your lemmatizer into spacy?

**Comment by Liebeck** [comment]
@smartinsightsfromdata I haven't been able to find time yet. Currently, there are only two major to-dos that I need to complete first, but I still plan to integrate IWNLP as a spaCy extension ;-)

**Comment by Liebeck** [comment]
I'm now starting to work on the extension.

By looking at https://spacy.io/usage/resources#extensions and https://github.com/topics/spacy-extension?o=desc&s=stars I see a couple of extension with different naming conventions. 

I'm either going for `spacy_iwnlp` or `spacy-iwnlp` for the pip package. @ines  @honnibal Would you mind making an official suggestion as to use a dash or an underscore? ;-)

**Comment by Liebeck** [comment]
@smartinsightsfromdata I just released https://github.com/Liebeck/spacy-iwnlp  

`pip install spacy-iwnlp`
